ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java

```diff
@@ -112,15 +145,59 @@ protected int getPartitions(final KafkaTopicClient kafkaTopicClient) {
     return right.getPartitions(kafkaTopicClient);
   }
 
+  @SuppressWarnings("UnstableApiUsage")
   @Override
   public Stream<ColumnName> resolveSelectStar(
-      final Optional<SourceName> sourceName, final boolean valueOnly
+      final Optional<SourceName> sourceName,
+      final boolean valueOnly
   ) {
-    return getSources()
-            .stream()
-            .flatMap(s -> s instanceof JoinNode ? s.getSources().stream() : Stream.of(s))
-            .filter(s -> !sourceName.isPresent() || sourceName.equals(s.getSourceName()))
-            .flatMap(s -> s.resolveSelectStar(sourceName, false));
+    final Stream<ColumnName> names = Stream.of(left, right)
+        .flatMap(s -> s instanceof JoinNode ? s.getSources().stream() : Stream.of(s))
+        .filter(s -> !sourceName.isPresent() || sourceName.equals(s.getSourceName()))
+        .flatMap(s -> s.resolveSelectStar(sourceName, false));
+
+    if (sourceName.isPresent() || joinKey.getType() != Type.SYNTHETIC || !finalJoin) {
+      return names;
+    }
+
+    final Column syntheticKey = Iterables.getOnlyElement(getSchema().key());
+
+    return Streams.concat(Stream.of(syntheticKey.name()), names);
+  }
+
+  public Expression resolveSelect(final int idx, final Expression expression) {
+    return joinKey.resolveSelect(expression, getSchema());
+  }
+
+  @SuppressWarnings("UnstableApiUsage")
+  @Override
+  void validateKeyPresent(final SourceName sinkName, final Projection projection) {
+
+    final boolean atLeastOneKey = Streams.concat(
+        joinKey.getAvailableKeys().stream(),
+        joinKey.getOriginalAvailableKeys().stream()
+    ).anyMatch(projection::containsExpression);
+
+    if (!atLeastOneKey) {
```

Comment:
With this PR, yes, it defaults the key column to the left join column, though you're right in that it could be the right join column.

In the future, both key columns should be in the key.